### DISCUSSION TITLE
Add a backtest module?

### DISCUSSION THREAD

**Discussion by kayahasa** [DaveSkender/Stock.Indicators#484] [Ideas]
It would be nice to have a backtest module

**Comment by DaveSkender** [comment]
I really like the idea of having a backtest engine that is affiliated with the library, though I would build it as an entirely separate project because a) it's a better architectural design pattern to separate responsibilities, and 2) indicator users won't always want our backtest engine bundled with the indicators.

Some ideas and basic design parameters I've been considering:

- Entirely separate package, as mentioned above
- Comes with basic strategies (e.g. crossovers, limits, etc)
- Some ability to build and use user custom strategies, since there's really no finite number of strategies
- Limited to signal generation (long vs short), initially
- Later, add optional full paper trade and account simulation (e.g. start with $10K, trades with BTO, STO, BTC, STC and cash/margin/fees management, ending balance, etc.); though, this paper trade part might be another project.

My biggest concerns are about modularity and whether we can maintain a finite scope of it and have it still be useful.  For example, once you have signals, how much analysis and presentation would you want ... charts ... live trading, etc?  It can grow into a full algo trading system if not careful, which is something I'd like to avoid (for this proposed module).

**Reply by DaveSkender** [reply]
As a side note, I've also been thinking about how to do basic pure analysis on individual indicators (not trade strategies), but that's somewhat different than a full strategy backtest engine.  See #320

**Reply by jimtollan** [reply]
Hi Dave - thanks for the library, goes without saying that it is a gamechanger as well as a real timesaver. The wheel that i was busy re-inventing, now put to the side.!!

Now onto the second wheel :D. Signal generation from the indicators library. I really love the idea of having a *plugin/related* library that uses the indicators and produces the signals when we have crossovers etc. One thing that really floats my boat is the thought of a signal being generated when a specific arrangement of candles is present (i.e. something that recognises doji reversals, bullish hammer, etc, etc). My expertise and experience are justso lacking as to make this a daunting, if not impossible task.

Would definitely be a gamechanger!!

**Reply by DaveSkender** [reply]
See also our conversation on candlestick patterns #220.  We still have quite a few of these to implement.

**Reply by jimtollan** [reply]
dave - thanks for the speedy response. i will hop over and have a look at the candlestick discussion you link. that could well be the answer to 80% of my current needs... have a grt wknd!

**Comment by DaveSkender** [comment]
I was just freshening up my thinking on backtests and all of the overfitting and problems to avoid.  It can be a real misleading mess if not careful.

What I’m reading:
[Backtest overfitting and the post-hoc probability fallacy](https://mathinvestor.org/2022/02/backtest-overfitting-and-the-post-hoc-probability-fallacy/) by David H Bailey

**Reply by tiger2014** [reply]
It will be better if this backtest project includes a chart to show the candles and the buy/sell signal. We can intuitively know how our strategies are.

If this will not include a chart, it might not be very helpful for us. You know, writing a backtest project without a chart is very simple for a backend developer (I think most people reading this project is a backend developers).

**Reply by DaveSkender** [reply]
@tiger2014 chart idea: https://github.com/DaveSkender/Stock.Indicators/pull/1021
